### ISSUE TITLE
Flag a output as "Safe" when there is the persistent tag to avoid dvc unprotect from being run

### ISSUE THREAD

**Issue by menegais1** [iterative/dvc#6562]
Due to memory concerns, some of my pipeline stages have persistent outputs that I handle in some python scripts. The pipeline stage receives a JSON as input and outputs a folder containing image files. As there is no need to rewrite those files every execution, I check inside the folder if a file is present (read-only) to avoid reprocessing it. As the dataset can get quite large, the space consumption becomes worrisome, as every **dvc repro** executed needs to unprotect all the files in the folder, copying them from the cache to the workspace. If an output could be marked as safe, so it would only suffer from append/remove operations, the unprotect could be avoided, reducing the space usage.

**Comment by shcheklein** [comment]
Probably we can even avoid running unprotect in this at all? Or unprotect dir - make it writable. Seem like there should be a config option for this.

**Comment by shcheklein** [comment]
Thinking even more about this, I've renamed the ticket a bit.

To be honest, it's not that clear why do we do this extra step (unprotect) in the first place. There are a lot of cases (majority of them to be honest that come to my mind) is when we need an append semantics (script doesn't overwrite the files). It can be very expensive to unprotect things and protect them back every time. 

I'm not sure about checkpoints though, but we also use a different flag for them for other reasons?

cc @dberenbaum @pmrowla

**Comment by dberenbaum** [comment]
For checkpoints, I think it will be necessary to unprotect because it's about loading state from the existing model file and writing an updated model file at the end. However, configuring a flag for `dvc repro` and `dvc exp run` generally to avoid needing to unprotect is probably fine.

**Comment by efiop** [comment]
@dberenbaum Just to confirm, you meant that we would need a dvc.yaml option for each such output, right? Similar to what is implemented in https://github.com/iterative/dvc/pull/7072  by @rgferrari ?

**Comment by dberenbaum** [comment]
Yes, this was a mistake by me. I was thinking of a `dvc.yaml` option since it seems more useful to permanently specify the behavior than have to include the option every time you reproduce the pipeline.

**Comment by uditrana** [comment]
Hi all, I am here to re-inject some interest in this issue ([relevant discord thread](https://discord.com/channels/485586884165107732/1098660228939124817)). 

I am working with a pretty common workflow where there is a "prep_data.py" script that creates cleaned data for training and other downstream stages. 

The problem is that the scenario involves a shared-cache (multiple users on same machine with a shared data drive). Any sort of pipeline operation (e.g. `dvc repro`) involves every file created by to be unprotected which is extremely expensive. The 'prep_data.py' script always deletes and recreates new files if there are detected changes so I am not worried about corruption (that much). 

I think an option to not unprotect and/or allowing us to programatically control projection (via python sdk) would be useful

**Comment by dberenbaum** [comment]
> To be honest, it's not that clear why do we do this extra step (unprotect) in the first place. There are a lot of cases (majority of them to be honest that come to my mind) is when we need an append semantics (script doesn't overwrite the files). It can be very expensive to unprotect things and protect them back every time.
> 
> I'm not sure about checkpoints though, but we also use a different flag for them for other reasons?

Now that we don't have checkpoints, should we consider dropping unprotect as default behavior completely? I guess we aren't in a good position to make this kind of breaking change right now, but at least if we introduce the flag soon, we could make it the default in the next major release.

**Comment by emilhotkowski** [comment]
@dberenbaum Hi, is this still something you want to add? I wanted to start contributing to DVC and it seems like a nice starter task :)
